jobs/pbm_helpers.go

```diff
@@ -0,0 +1,334 @@
+		Type    string `json:"type"`
+		Name    string `json:"name"`
+		StartTS int    `json:"startTS"`
+		Status  string `json:"status"`
+		OpID    string `json:"opID"`
+	} `json:"running"`
+}
+
+func getPBMOutput(ctx context.Context, dbURL *url.URL, to interface{}, args ...string) error {
+	nCtx, cancel := context.WithTimeout(ctx, cmdTimeout)
+	defer cancel()
+
+	args = append(args, "--out=json", "--mongodb-uri="+dbURL.String())
+	cmd := exec.CommandContext(nCtx, pbmBin, args...) // #nosec G204
+
+	b, err := cmd.Output()
+	if err != nil {
+		var exitErr *exec.ExitError
+		if errors.As(err, &exitErr) {
+			return errors.New(string(exitErr.Stderr))
+		}
+		return err
+	}
+
+	return json.Unmarshal(b, to)
+}
+
+func pbmSetupS3(ctx context.Context, l logrus.FieldLogger, dbURL *url.URL, prefix string, s3Config *S3LocationConfig, resync bool) error {
+	l.Info("Configuring S3 location.")
+	nCtx, cancel := context.WithTimeout(ctx, cmdTimeout)
+	defer cancel()
+
+	confFile, err := writePBMConfigFile(prefix, s3Config)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	defer os.Remove(confFile) //nolint:errcheck
+
+	output, err := exec.CommandContext( //nolint:gosec
+		nCtx,
+		pbmBin,
+		"config",
+		"--mongodb-uri="+dbURL.String(),
+		"--file="+confFile,
+	).CombinedOutput()
+
+	if err != nil {
+		return errors.Wrapf(err, "pbm config error: %s", string(output))
+	}
+
+	if resync {
+		nCtx, cancel := context.WithTimeout(ctx, cmdTimeout)
+		defer cancel()
+
+		output, err = exec.CommandContext( //nolint:gosec
+			nCtx,
+			pbmBin,
+			"config",
+			"--mongodb-uri="+dbURL.String(),
+			"--force-resync",
+		).CombinedOutput()
+
+		if err != nil {
+			return errors.Wrapf(err, "pbm config error: %s", string(output))
+		}
+	}
+
+	return nil
+}
+
+func retrieveLogs(ctx context.Context, dbURL *url.URL, event string) ([]pbmLogEntry, error) {
+	var logs []pbmLogEntry
+
+	if err := getPBMOutput(ctx, dbURL, &logs, "logs", "--event="+event, "--tail=0"); err != nil {
+		return nil, err
+	}
+
+	return logs, nil
+}
+
+type pbmStatusCondition func(s pbmStatus) (bool, error)
+
+func pbmNoRunningOperations(s pbmStatus) (bool, error) {
+	return s.Running.Status == "", nil
+}
+
+func pbmBackupFinished(name string) pbmStatusCondition {
+	started := false
+	checks := 0
+	return func(s pbmStatus) (bool, error) {
+		checks++
+		if s.Running.Type == "backup" && s.Running.Name == name && s.Running.Status != "" {
+			started = true
+		}
+		if !started && checks > maxBackupChecks {
+			return false, errors.New("failed to start backup")
+		}
+		var snapshot *pbmSnapshot
+		for i, snap := range s.Backups.Snapshot {
+			if snap.Name == name {
+				snapshot = &s.Backups.Snapshot[i]
+				break
+			}
+		}
+		if snapshot == nil || s.Running.Status != "" {
+			return false, nil
+		}
+
+		if snapshot.Status == "error" {
+			return false, errors.New(snapshot.Error)
+		}
+		return snapshot.Status == "done", nil
+	}
+}
+
+func pbmRestoreFinished(name string) pbmStatusCondition {
```

Comment:
this will be rewritten to check output from `pbm list --restore` instead of relying on running operation